src/Workleap.DomainEventPropagation.Subscription.PullDelivery/EventPuller.cs

```diff
@@ -0,0 +1,88 @@
+﻿using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Workleap.DomainEventPropagation.EventGridClientAdapter;
+
+namespace Workleap.DomainEventPropagation;
+
+internal class EventPuller : BackgroundService
+{
+    private readonly IServiceScopeFactory _serviceScopeFactory;
+    private readonly ILogger<EventPuller> _logger;
+    private readonly EventGridTopicSubscription[] _eventGridTopicSubscriptions;
+
+    public EventPuller(
+        IServiceScopeFactory serviceScopeFactory,
+        IEnumerable<EventGridClientDescriptor> clientDescriptors,
+        IEventGridClientWrapperFactory eventGridClientWrapperFactory,
+        IOptionsMonitor<EventPropagationSubscriptionOptions> optionsMonitor,
+        ILogger<EventPuller> logger)
+    {
+        this._serviceScopeFactory = serviceScopeFactory;
+        this._logger = logger;
+        this._eventGridTopicSubscriptions = clientDescriptors.Select(descriptor =>
+            new EventGridTopicSubscription(
+                optionsMonitor.Get(descriptor.Name).TopicName,
+                optionsMonitor.Get(descriptor.Name).SubscriptionName,
+                eventGridClientWrapperFactory.CreateClient(descriptor.Name))).ToArray();
+    }
+
+    protected override Task ExecuteAsync(CancellationToken stoppingToken)
+    {
+        return Task.WhenAll(this._eventGridTopicSubscriptions.Select(sub => Task.Run(() => this.StartReceivingEventsAsync(sub, this._logger, stoppingToken), stoppingToken)));
+    }
+
+    private async Task StartReceivingEventsAsync(EventGridTopicSubscription eventGridTopicSubscription, ILogger<EventPuller> logger, CancellationToken stoppingToken)
+    {
+        using var scope = this._serviceScopeFactory.CreateScope();
+        var cloudEventHandler = scope.ServiceProvider.GetRequiredService<ICloudEventHandler>();
+
+        while (!stoppingToken.IsCancellationRequested)
+        {
+            try
+            {
+                var bundles = await eventGridTopicSubscription.Client.ReceiveCloudEventsAsync(eventGridTopicSubscription.TopicName, eventGridTopicSubscription.SubscriptionName, cancellationToken: stoppingToken).ConfigureAwait(false);
+                foreach (var (cloudEvent, lockToken) in bundles)
+                {
+                    var status = await cloudEventHandler.HandleCloudEventAsync(cloudEvent, stoppingToken).ConfigureAwait(false);
```

Comment:
Seriez-vous open à avoir le cloudEventHandler lancer des exceptions qui seraient handler ici, je crois que ça permettrait d'avoir les deux flows de subscription (push et pull) utiliser la même signature de pipeline et potentiellement bénéficier de la telemetrie plus facilement. Je serais partant pour en jaser si ça vous dit

Reply:
As discussed, we will put this aside for the time being and convert it to tech debt that might be addressed later on.